cpp/include/cudf/detail/aggregation/aggregation.hpp

```diff
@@ -642,50 +639,69 @@ class rank_aggregation final : public rolling_aggregation,
+                   bool percentage)
+    : aggregation{RANK},
+      _method{method},
+      _column_order{column_order},
+      _null_handling{null_handling},
+      _null_precedence{null_precedence},
+      _percentage(percentage)
   {
-    return std::make_unique<rank_aggregation>(*this);
   }
-  std::vector<std::unique_ptr<aggregation>> get_simple_aggregations(
-    data_type col_type, simple_aggregations_collector& collector) const override
+  rank_method const _method;          ///< rank method
+  order const _column_order;          ///< order of the column to rank
+  null_policy const _null_handling;   ///< include or exclude nulls in ranks
+  null_order const _null_precedence;  ///< order of nulls in ranks
+  bool const _percentage;             ///< whether to return percentage ranks
+
+  [[nodiscard]] bool is_equal(aggregation const& _other) const override
   {
-    return collector.visit(col_type, *this);
+    if (!this->aggregation::is_equal(_other)) { return false; }
+    auto const& other = dynamic_cast<rank_aggregation const&>(_other);
+    return _method == other._method and _null_handling == other._null_handling and
+           _column_order == other._column_order and _null_precedence == other._null_precedence and
+           _percentage == other._percentage;
   }
-  void finalize(aggregation_finalizer& finalizer) const override { finalizer.visit(*this); }
-};
 
-/**
- * @brief Derived class for specifying a dense rank aggregation
- */
-class dense_rank_aggregation final : public rolling_aggregation,
-                                     public groupby_scan_aggregation,
-                                     public scan_aggregation {
- public:
-  dense_rank_aggregation() : aggregation{DENSE_RANK} {}
+  [[nodiscard]] size_t do_hash() const override
+  {
+    return this->aggregation::do_hash() ^ hash_impl();
+  }
 
   [[nodiscard]] std::unique_ptr<aggregation> clone() const override
   {
-    return std::make_unique<dense_rank_aggregation>(*this);
+    return std::make_unique<rank_aggregation>(*this);
   }
   std::vector<std::unique_ptr<aggregation>> get_simple_aggregations(
     data_type col_type, simple_aggregations_collector& collector) const override
   {
     return collector.visit(col_type, *this);
   }
   void finalize(aggregation_finalizer& finalizer) const override { finalizer.visit(*this); }
+
+ private:
+  [[nodiscard]] size_t hash_impl() const
+  {
+    return std::hash<int>{}(static_cast<int>(_method)) ^
+           std::hash<int>{}(static_cast<int>(_column_order)) ^
+           std::hash<int>{}(static_cast<int>(_null_handling)) ^
+           std::hash<int>{}(static_cast<int>(_null_precedence)) ^ std::hash<bool>{}(_percentage);
+  }
 };
 
-class percent_rank_aggregation final : public rolling_aggregation,
-                                       public groupby_scan_aggregation,
-                                       public scan_aggregation {
+class ansi_sql_percent_rank_aggregation final : public rolling_aggregation,
```

Comment:
Why is this a distinct aggregation type instead of an argument to the normal `rank_aggregation`?

Reply:
Refer https://github.com/rapidsai/cudf/pull/9569#issuecomment-1050388844

> 3. Keep `PERCENT_RANK` as a separate aggregation. (Explained below.)
> 
> #3 above warrants justification, and discussion. On first glance, it would appear that ANSI SQL's `PERCENT_RANK` semantics could be addressed similarly to Pandas's `Dataframe.rank(method='MIN', pct='True')`. I suspect it can't, because:
> 
> ```
> pandas_min_percent_rank == (row_rank / num_rows_in_group);
> sql_percent_rank == ((row_rank - 1) / (num_rows_in_group-1));
> ```
> 
> Also, there is a 1-1 correspondence between `cudf::rank_method` and Pandas's `Dataframe.rank(method)`. It would be erroneous/nonsensical to attempt `PERCENT_RANK` aggregations with `pct=True`. My vote is currently to leave `PERCENT_RANK` separate. There might be value in renaming `PERCENT_RANK` to `ANSI_SQL_PERCENT_RANK`, to clarify that this is ANSI SQL compliant.

Reply:
@karthikeyann @mythrocks that doesn't sound like sufficient justification for a distinct aggregation when the only difference is a slight change in how the "percentage" is calculated. 

For example, look at the standard deviation aggregation and how it exposes the `ddof` parameter to control a constant offset to the population size: https://github.com/rapidsai/cudf/blob/565f4743f797ff31afd402f715a4c57547cb6c66/cpp/include/cudf/aggregation.hpp#L252-L261

Instead of making the `percentage` argument to `make_rank_aggregation` be a `bool`, make it something where you can expose control over how the percentage is calculated.

Do *not* call it `sql` and `pandas` methods. It should be a red flag if you have to refer to a specific framework for naming something. It should just be described in terms of the math.

Reply:
To put in plain words about the difference between `sql` and `pandas` percentage, 
`sql` percentage rank ranges from 0% to 100% (only applies for MIN method, all other methods are not applicable).
`pandas` percentage rank ranges from >0% to 100%. (all methods are applicable)

Reply:
To describe in terms of math, how about adding new method named `0-indexed-min-rank` and pass `percentage` true instead of `ANSI_SQL_PERCENT_RANK`? (but `percentage` = false will not be supported.)

Note that We are dividing by maximum of group rank value (which is group_size-1 ), not the group size.

Reply:
@jrhemstad replaced `ANSI_SQL_PERCENT_RANK` aggregation with `MIN_0_INDEXED` rank_method
commit [c988d8f](https://github.com/rapidsai/cudf/pull/9569/commits/c988d8f8244476c1a3954410440e64282415a426)

Reply:
@karthikeyann we had a longer discussion of this issue at the libcudf team meeting today to get aligned on what we want this API to look like. The current `rank_method`s aside from the new `MIN_0_INDEXED` are all about how ties are broken, which is very different from whether we compute a percentage or not. Our proposal is to instead use a separate enum like
```
enum class rank_percentage : int32_t {
  NONE,                                ///< rank
  ZERO_NORMALIZED,        ///< rank / group_size
  ONE_NORMALIZED           ///< (rank - 1) / (group_size - 1)
};
```
so that control over the percentage or not part of the calculation is independent of the tiebreaking. @mythrocks was tentatively supportive of this, but he mentioned that there was potentially a technical blocker here because getting the `group_size` inside the aggregation was a problem. Could you could identify and expand upon that problem? Figuring out where we have a problem there would help us move towards a better consensus solution.

Reply:
updated code with the discussion solution. Added `rank_percentage`.